design-proposals/migration-target.md

```diff
@@ -0,0 +1,260 @@
+
+## Repos
+- https://github.com/kubevirt/kubevirt
+
+# Design
+## Proposed design
+We are going to add a new optional `addedNodeSelectorTerm` stanza of type `*k8sv1.NodeSelectorTerm` on the `VirtualMachineInstanceMigration` object.
+We are not going to alter by any mean the `spec` stanza of the VM or the VMI objects so future migrations or the node placement after a restart of the VM are not going to be affected by a `addedNodeSelectorTerm` set on a specific `VirtualMachineInstanceMigration` object.
+If the target pod fails to be started, the `VirtualMachineInstanceMigration` object will be marked as failed as it can already happen today for other reasons.
+The reason of the eventual failure will be reported back as it gets reported back today when a migration fails due to other reasons. 
+
+## Why addedNodeSelectorTerm and not simply addedNodeSelector
+
+According to the [k8s APIs](https://github.com/kubernetes/api/blob/71385f038c1097af36f3d2f68b415860b866c1f8/core/v1/types.go#L3355-L3363), a `nodeSelector` is a list of `NodeSelectorTerms` and
+*it represents the OR of the selectors represented by the node selector terms*.
+This means that a Pod can be scheduled onto a node if just one of the specified `NodeSelectorTerms` can be satisfied (terms are ORed).
+This means that if a catch all `NodeSelectorTerm` is added in addition to existing `NodeSelectorTerms` already defined at VM level, it will completely defeat and bypass the constraints defined by the VM owner on the VM object while this proposal is only about being able to restrict the set of valid target nodes for a migration adding additional constraints (`pure AND logic`).
+In k8s APIs, `NodeSelectorTerm` are [ORed](https://github.com/kubernetes/api/blob/71385f038c1097af36f3d2f68b415860b866c1f8/core/v1/types.go#L3360) while `NodeSelectorRequirements` within a single `NodeSelectorTerm` are [ANDed](https://github.com/kubernetes/api/blob/71385f038c1097af36f3d2f68b415860b866c1f8/core/v1/types.go#L3366).
+This proposal is only about exposing **pure AND logic** to limit the set of candidate nodes for a live migration still respecting what is specified on the VM object so exposing the `NodeSelector` API is not an option.
+Exposing a single `NodeSelectorTerm` on the `VMIM` object and adding all of the `NodeSelectorRequirements` defined there to all of the `NodeSelectorTerms` already defined on the VM object is instead a viable solution to achieve *pure AND logic*.
+
+### Why to simply `NodeSelector map[string]string`
+On Pod spec we also have `NodeSelector map[string]string` which is used in **AND** with `NodeAffinity` rules so from this point of view its a viable option.
+On the other side `pod.spec.nodeSelector` is only matching labels and the predefined `kubernetes.io/hostname` [label is not guaranteed to be reliable](https://kubernetes.io/docs/reference/node/node-labels/#preset-labels).
```

Comment:
Looking at this closer, I think using NodeSelector is the better option.

The issues I have with NodeSelectorTerm mostly have to do with how awkward it is to expose this tunable outside of the Affinity struct. The NodeSelectorTerm isn't really it's own concept exactly. It's one part of the pod Affinity and it's also a part of Affinity that has different meanings depending if it's in the preferred or required during scheduling section of the NodeAffinity. I'd rather not have to expose to users all these details.

While NodeSelector is less expressive, i think it's the most straightforward for UX for users, and i think it satisfies the goals here.